feat(components): [autocomplete] support home and end key events

snowbitx opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

This PR adds support for the Home and End keys in the keydown event, and consolidates the logic into the script to make event handling easier to maintain.

snowbitx committed:
feat(components): [autocomplete] support home and end key events

snowbitx commented:
![PixPin_2025-09-25_23-56-31](https://github.com/user-attachments/assets/99546b4c-5b73-4d36-9afc-41adef49b648)

FrontEndDog commented:
I think `PgUp` and `PgDn` are used more than `Home` and `End`. Can you add them together.
At the same time, I discovered a bug.
[link](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2IGNsYXNzPVwiZGVtby1hdXRvY29tcGxldGVcIj5cbiAgICA8ZGl2IGNsYXNzPVwiZGVtby1ibG9ja1wiPlxuICAgICAgPGRpdiBjbGFzcz1cImRlbW8tdGl0bGVcIj5saXN0IHN1Z2dlc3Rpb25zIHdoZW4gYWN0aXZhdGVkPC9kaXY+XG4gICAgICA8ZWwtYXV0b2NvbXBsZXRlXG4gICAgICAgIHYtbW9kZWw9XCJzdGF0ZTFcIlxuICAgICAgICA6ZmV0Y2gtc3VnZ2VzdGlvbnM9XCJxdWVyeVNlYXJjaFwiXG4gICAgICAgIGNsZWFyYWJsZVxuICAgICAgICBjbGFzcz1cInctNTBcIlxuICAgICAgICBwbGFjZWhvbGRlcj1cIlBsZWFzZSBJbnB1dFwiXG4gICAgICAgIEBzZWxlY3Q9XCJoYW5kbGVTZWxlY3RcIlxuICAgICAgLz5cbiAgICA8L2Rpdj5cbiAgICA8ZGl2IGNsYXNzPVwiZGVtby1ibG9ja1wiPlxuICAgICAgPGRpdiBjbGFzcz1cImRlbW8tdGl0bGVcIj5saXN0IHN1Z2dlc3Rpb25zIG9uIGlucHV0PC9kaXY+XG4gICAgICA8ZWwtYXV0b2NvbXBsZXRlXG4gICAgICAgIHYtbW9kZWw9XCJzdGF0ZTJcIlxuICAgICAgICA6ZmV0Y2gtc3VnZ2VzdGlvbnM9XCJxdWVyeVNlYXJjaFwiXG4gICAgICAgIDp0cmlnZ2VyLW9uLWZvY3VzPVwiZmFsc2VcIlxuICAgICAgICBjbGVhcmFibGVcbiAgICAgICAgY2xhc3M9XCJ3LTUwXCJcbiAgICAgICAgcGxhY2Vob2xkZXI9XCJQbGVhc2UgSW5wdXRcIlxuICAgICAgICBAc2VsZWN0PVwiaGFuZGxlU2VsZWN0XCJcbiAgICAgIC8+XG4gICAgPC9kaXY+XG4gIDwvZGl2PlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdCBsYW5nPVwidHNcIiBzZXR1cD5cbmltcG9ydCB7IG9uTW91bnRlZCwgcmVmIH0gZnJvbSAndnVlJ1xuXG5pbnRlcmZhY2UgUmVzdGF1cmFudEl0ZW0ge1xuICB2YWx1ZTogc3RyaW5nXG4gIGxpbms6IHN0cmluZ1xufVxuXG5jb25zdCBzdGF0ZTEgPSByZWYoJycpXG5jb25zdCBzdGF0ZTIgPSByZWYoJycpXG5cbmNvbnN0IHJlc3RhdXJhbnRzID0gcmVmPFJlc3RhdXJhbnRJdGVtW10+KFtdKVxuY29uc3QgcXVlcnlTZWFyY2ggPSAocXVlcnlTdHJpbmc6IHN0cmluZywgY2I6IGFueSkgPT4ge1xuICBjb25zdCByZXN1bHRzID0gcXVlcnlTdHJpbmdcbiAgICA/IHJlc3RhdXJhbnRzLnZhbHVlLmZpbHRlcihjcmVhdGVGaWx0ZXIocXVlcnlTdHJpbmcpKVxuICAgIDogcmVzdGF1cmFudHMudmFsdWVcbiAgLy8gY2FsbCBjYWxsYmFjayBmdW5jdGlvbiB0byByZXR1cm4gc3VnZ2VzdGlvbnNcbiAgY2IocmVzdWx0cylcbn1cbmNvbnN0IGNyZWF0ZUZpbHRlciA9IChxdWVyeVN0cmluZzogc3RyaW5nKSA9PiB7XG4gIHJldHVybiAocmVzdGF1cmFudDogUmVzdGF1cmFudEl0ZW0pID0+IHtcbiAgICByZXR1cm4gKFxuICAgICAgcmVzdGF1cmFudC52YWx1ZS50b0xvd2VyQ2FzZSgpLmluZGV4T2YocXVlcnlTdHJpbmcudG9Mb3dlckNhc2UoKSkgPT09IDBcbiAgICApXG4gIH1cbn1cbmNvbnN0IGxvYWRBbGwgPSAoKSA9PiB7XG4gIHJldHVybiBbXG4gICAgeyB2YWx1ZTogJ3Z1ZScsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vdnVlanMvdnVlJyB9LFxuICAgIHsgdmFsdWU6ICdlbGVtZW50JywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS9FbGVtZUZFL2VsZW1lbnQnIH0sXG4gICAgeyB2YWx1ZTogJ2Nvb2tpbmcnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL0VsZW1lRkUvY29va2luZycgfSxcbiAgICB7IHZhbHVlOiAnbWludC11aScsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vRWxlbWVGRS9taW50LXVpJyB9LFxuICAgIHsgdmFsdWU6ICd2dWV4JywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS92dWVqcy92dWV4JyB9LFxuICAgIHsgdmFsdWU6ICd2dWUtcm91dGVyJywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS92dWVqcy92dWUtcm91dGVyJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDInLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDMnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDQnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDUnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDYnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDcnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDgnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDknLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDExMCcsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vYmFiZWwvYmFiZWwnIH0sXG4gICAgeyB2YWx1ZTogJ2JhYmVsMTEnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDEyJywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS9iYWJlbC9iYWJlbCcgfSxcbiAgICB7IHZhbHVlOiAnYmFiZWwzMScsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vYmFiZWwvYmFiZWwnIH0sXG4gICAgeyB2YWx1ZTogJ2JhYmVsMTQnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDE1JywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS9iYWJlbC9iYWJlbCcgfSxcbiAgICB7IHZhbHVlOiAnYmFiZWwxNicsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vYmFiZWwvYmFiZWwnIH0sXG4gIF1cbn1cblxuY29uc3QgaGFuZGxlU2VsZWN0ID0gKGl0ZW06IFJlY29yZDxzdHJpbmcsIGFueT4pID0+IHtcbiAgY29uc29sZS5sb2coaXRlbSlcbn1cblxub25Nb3VudGVkKCgpID0+IHtcbiAgcmVzdGF1cmFudHMudmFsdWUgPSBsb2FkQWxsKClcbn0pXG48L3NjcmlwdD5cblxuPHN0eWxlIHNjb3BlZD5cbi5kZW1vLWF1dG9jb21wbGV0ZSB7XG4gIGRpc3BsYXk6IGZsZXg7XG4gIGZsZXgtd3JhcDogd3JhcDtcbiAgZ2FwOiAycmVtO1xufVxuXG4uZGVtby1ibG9jayB7XG4gIGRpc3BsYXk6IGZsZXg7XG4gIGZsZXgtZGlyZWN0aW9uOiBjb2x1bW47XG4gIGdhcDogMC41cmVtO1xufVxuXG4uZGVtby10aXRsZSB7XG4gIGZvbnQtc2l6ZTogMC44NzVyZW07XG4gIGNvbG9yOiB2YXIoLS1lbC10ZXh0LWNvbG9yLXNlY29uZGFyeSk7XG4gIG1pbi1oZWlnaHQ6IDIuNWVtO1xuICBkaXNwbGF5OiBmbGV4O1xuICBhbGlnbi1pdGVtczogY2VudGVyO1xufVxuXG5AbWVkaWEgc2NyZWVuIGFuZCAobWF4LXdpZHRoOiA3NjhweCkge1xuICAuZGVtby1hdXRvY29tcGxldGUge1xuICAgIGdhcDogMXJlbTtcbiAgfVxuXG4gIC5kZW1vLWJsb2NrIHtcbiAgICB3aWR0aDogMTAwJTtcbiAgfVxufVxuPC9zdHlsZT5cbiIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzJywgJ2h0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwidW5zdXBwb3J0ZWRcIixcbiAgICBcIkBlbGVtZW50LXBsdXMvaWNvbnMtdnVlXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AZWxlbWVudC1wbHVzL2ljb25zLXZ1ZUAyL2Rpc3QvaW5kZXgubWluLmpzXCJcbiAgfSxcbiAgXCJzY29wZXNcIjoge31cbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzIn19)

When the scrollbar appears, Home and End exhibit unexpected behavior.

tolking commented:
This is a relatively reasonable behavior. [Link](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/)

snowbitx committed:
refactor: add pageUp+pageDown and fix scroll error

snowbitx committed:
Merge branch 'dev' into feat/autocomplete-keydown

snowbitx commented:
> I think `PgUp` and `PgDn` are used more than `Home` and `End`. Can you add them together. At the same time, I discovered a bug. [link](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2IGNsYXNzPVwiZGVtby1hdXRvY29tcGxldGVcIj5cbiAgICA8ZGl2IGNsYXNzPVwiZGVtby1ibG9ja1wiPlxuICAgICAgPGRpdiBjbGFzcz1cImRlbW8tdGl0bGVcIj5saXN0IHN1Z2dlc3Rpb25zIHdoZW4gYWN0aXZhdGVkPC9kaXY+XG4gICAgICA8ZWwtYXV0b2NvbXBsZXRlXG4gICAgICAgIHYtbW9kZWw9XCJzdGF0ZTFcIlxuICAgICAgICA6ZmV0Y2gtc3VnZ2VzdGlvbnM9XCJxdWVyeVNlYXJjaFwiXG4gICAgICAgIGNsZWFyYWJsZVxuICAgICAgICBjbGFzcz1cInctNTBcIlxuICAgICAgICBwbGFjZWhvbGRlcj1cIlBsZWFzZSBJbnB1dFwiXG4gICAgICAgIEBzZWxlY3Q9XCJoYW5kbGVTZWxlY3RcIlxuICAgICAgLz5cbiAgICA8L2Rpdj5cbiAgICA8ZGl2IGNsYXNzPVwiZGVtby1ibG9ja1wiPlxuICAgICAgPGRpdiBjbGFzcz1cImRlbW8tdGl0bGVcIj5saXN0IHN1Z2dlc3Rpb25zIG9uIGlucHV0PC9kaXY+XG4gICAgICA8ZWwtYXV0b2NvbXBsZXRlXG4gICAgICAgIHYtbW9kZWw9XCJzdGF0ZTJcIlxuICAgICAgICA6ZmV0Y2gtc3VnZ2VzdGlvbnM9XCJxdWVyeVNlYXJjaFwiXG4gICAgICAgIDp0cmlnZ2VyLW9uLWZvY3VzPVwiZmFsc2VcIlxuICAgICAgICBjbGVhcmFibGVcbiAgICAgICAgY2xhc3M9XCJ3LTUwXCJcbiAgICAgICAgcGxhY2Vob2xkZXI9XCJQbGVhc2UgSW5wdXRcIlxuICAgICAgICBAc2VsZWN0PVwiaGFuZGxlU2VsZWN0XCJcbiAgICAgIC8+XG4gICAgPC9kaXY+XG4gIDwvZGl2PlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdCBsYW5nPVwidHNcIiBzZXR1cD5cbmltcG9ydCB7IG9uTW91bnRlZCwgcmVmIH0gZnJvbSAndnVlJ1xuXG5pbnRlcmZhY2UgUmVzdGF1cmFudEl0ZW0ge1xuICB2YWx1ZTogc3RyaW5nXG4gIGxpbms6IHN0cmluZ1xufVxuXG5jb25zdCBzdGF0ZTEgPSByZWYoJycpXG5jb25zdCBzdGF0ZTIgPSByZWYoJycpXG5cbmNvbnN0IHJlc3RhdXJhbnRzID0gcmVmPFJlc3RhdXJhbnRJdGVtW10+KFtdKVxuY29uc3QgcXVlcnlTZWFyY2ggPSAocXVlcnlTdHJpbmc6IHN0cmluZywgY2I6IGFueSkgPT4ge1xuICBjb25zdCByZXN1bHRzID0gcXVlcnlTdHJpbmdcbiAgICA/IHJlc3RhdXJhbnRzLnZhbHVlLmZpbHRlcihjcmVhdGVGaWx0ZXIocXVlcnlTdHJpbmcpKVxuICAgIDogcmVzdGF1cmFudHMudmFsdWVcbiAgLy8gY2FsbCBjYWxsYmFjayBmdW5jdGlvbiB0byByZXR1cm4gc3VnZ2VzdGlvbnNcbiAgY2IocmVzdWx0cylcbn1cbmNvbnN0IGNyZWF0ZUZpbHRlciA9IChxdWVyeVN0cmluZzogc3RyaW5nKSA9PiB7XG4gIHJldHVybiAocmVzdGF1cmFudDogUmVzdGF1cmFudEl0ZW0pID0+IHtcbiAgICByZXR1cm4gKFxuICAgICAgcmVzdGF1cmFudC52YWx1ZS50b0xvd2VyQ2FzZSgpLmluZGV4T2YocXVlcnlTdHJpbmcudG9Mb3dlckNhc2UoKSkgPT09IDBcbiAgICApXG4gIH1cbn1cbmNvbnN0IGxvYWRBbGwgPSAoKSA9PiB7XG4gIHJldHVybiBbXG4gICAgeyB2YWx1ZTogJ3Z1ZScsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vdnVlanMvdnVlJyB9LFxuICAgIHsgdmFsdWU6ICdlbGVtZW50JywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS9FbGVtZUZFL2VsZW1lbnQnIH0sXG4gICAgeyB2YWx1ZTogJ2Nvb2tpbmcnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL0VsZW1lRkUvY29va2luZycgfSxcbiAgICB7IHZhbHVlOiAnbWludC11aScsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vRWxlbWVGRS9taW50LXVpJyB9LFxuICAgIHsgdmFsdWU6ICd2dWV4JywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS92dWVqcy92dWV4JyB9LFxuICAgIHsgdmFsdWU6ICd2dWUtcm91dGVyJywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS92dWVqcy92dWUtcm91dGVyJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDInLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDMnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDQnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDUnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDYnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDcnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDgnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDknLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDExMCcsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vYmFiZWwvYmFiZWwnIH0sXG4gICAgeyB2YWx1ZTogJ2JhYmVsMTEnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDEyJywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS9iYWJlbC9iYWJlbCcgfSxcbiAgICB7IHZhbHVlOiAnYmFiZWwzMScsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vYmFiZWwvYmFiZWwnIH0sXG4gICAgeyB2YWx1ZTogJ2JhYmVsMTQnLCBsaW5rOiAnaHR0cHM6Ly9naXRodWIuY29tL2JhYmVsL2JhYmVsJyB9LFxuICAgIHsgdmFsdWU6ICdiYWJlbDE1JywgbGluazogJ2h0dHBzOi8vZ2l0aHViLmNvbS9iYWJlbC9iYWJlbCcgfSxcbiAgICB7IHZhbHVlOiAnYmFiZWwxNicsIGxpbms6ICdodHRwczovL2dpdGh1Yi5jb20vYmFiZWwvYmFiZWwnIH0sXG4gIF1cbn1cblxuY29uc3QgaGFuZGxlU2VsZWN0ID0gKGl0ZW06IFJlY29yZDxzdHJpbmcsIGFueT4pID0+IHtcbiAgY29uc29sZS5sb2coaXRlbSlcbn1cblxub25Nb3VudGVkKCgpID0+IHtcbiAgcmVzdGF1cmFudHMudmFsdWUgPSBsb2FkQWxsKClcbn0pXG48L3NjcmlwdD5cblxuPHN0eWxlIHNjb3BlZD5cbi5kZW1vLWF1dG9jb21wbGV0ZSB7XG4gIGRpc3BsYXk6IGZsZXg7XG4gIGZsZXgtd3JhcDogd3JhcDtcbiAgZ2FwOiAycmVtO1xufVxuXG4uZGVtby1ibG9jayB7XG4gIGRpc3BsYXk6IGZsZXg7XG4gIGZsZXgtZGlyZWN0aW9uOiBjb2x1bW47XG4gIGdhcDogMC41cmVtO1xufVxuXG4uZGVtby10aXRsZSB7XG4gIGZvbnQtc2l6ZTogMC44NzVyZW07XG4gIGNvbG9yOiB2YXIoLS1lbC10ZXh0LWNvbG9yLXNlY29uZGFyeSk7XG4gIG1pbi1oZWlnaHQ6IDIuNWVtO1xuICBkaXNwbGF5OiBmbGV4O1xuICBhbGlnbi1pdGVtczogY2VudGVyO1xufVxuXG5AbWVkaWEgc2NyZWVuIGFuZCAobWF4LXdpZHRoOiA3NjhweCkge1xuICAuZGVtby1hdXRvY29tcGxldGUge1xuICAgIGdhcDogMXJlbTtcbiAgfVxuXG4gIC5kZW1vLWJsb2NrIHtcbiAgICB3aWR0aDogMTAwJTtcbiAgfVxufVxuPC9zdHlsZT5cbiIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzJywgJ2h0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwidW5zdXBwb3J0ZWRcIixcbiAgICBcIkBlbGVtZW50LXBsdXMvaWNvbnMtdnVlXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AZWxlbWVudC1wbHVzL2ljb25zLXZ1ZUAyL2Rpc3QvaW5kZXgubWluLmpzXCJcbiAgfSxcbiAgXCJzY29wZXNcIjoge31cbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy0yMjMyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzIn19)
> 
> When the scrollbar appears, Home and End exhibit unexpected behavior.

Thanks for the feedback! Fixed the bug and added `PgUp` and `PgDn`

snowbitx commented:
> This is a relatively reasonable behavior. [Link](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/)

Thanks for the doc, I’ve adjusted the behavior accordingly.

cszhjh commented on packages/components/autocomplete/src/autocomplete.vue:
```diff
@@ -358,35 +355,90 @@ const highlight = (index: number) => {
+      highlight(calcPageIndex(1))
+    },
+  }
+  keyHandlers[e.key]?.()
```
Perhaps we can refer to #22228, use `getEventCode` to obtain KeyboardEvent.code.

https://github.com/element-plus/element-plus/blob/0e8b01787419c4343c405e673df16c7092849c13/packages/utils/dom/event.ts#L25

snowbitx committed:
chore: refer to #22228

snowbitx committed:
Merge branch 'feat/autocomplete-keydown' of https://github.com/snowbitx/element-plus into feat/autocomplete-keydown

snowbitx committed:
chore: change switch case

snowbitx committed:
Merge branch 'dev' into feat/autocomplete-keydown

snowbitx committed:
refactor: change steps 10

snowbitx committed:
Merge branch 'feat/autocomplete-keydown' of https://github.com/snowbitx/element-plus into feat/autocomplete-keydown

snowbitx commented on packages/components/autocomplete/src/autocomplete.vue:
```diff
@@ -358,35 +355,90 @@ const highlight = (index: number) => {
+  return direction === 1
+    ? Math.min(target, suggestions.value.length - 1)
+    : Math.max(target, 0)
+}
```
> [pageUp/pageDown] steps can be set to any value greater than 1, I think fixed 10 is more appropriate

Hey! My first version actually did that, but the PageUp/PageDown continuity didn’t feel quite right. Now I updated the PR to use a fixed step of 10. Please give it a try — if it needs further adjustment, just let me know. 🙂

tolking commented on packages/components/autocomplete/__tests__/autocomplete.test.tsx:
```diff
@@ -380,14 +381,14 @@ describe('Autocomplete.vue', () => {
 
     for (let i = 0; i < length; i++) {
-      await input.trigger('keydown.down')
+      await input.trigger('keydown', { key: EVENT_CODE.down })
```
```suggestion
      await input.trigger('keydown', { code: EVENT_CODE.down })
```

tolking commented on packages/components/autocomplete/__tests__/autocomplete.test.tsx:
```diff
@@ -406,18 +407,18 @@ describe('Autocomplete.vue', () => {
     }
 
-    await input.trigger('keydown.down')
+    await input.trigger('keydown', { key: EVENT_CODE.down })
```
Can you add some tests about the `home` `end` `pageUp` `pageUp` keys?

snowbitx replied:
> Can you add some tests about the `home` `end` `pageUp` `pageUp` keys?

Updated, thanks!

snowbitx committed:
Update packages/components/autocomplete/__tests__/autocomplete.test.tsx

Co-authored-by: qiang <qw13131wang@gmail.com>

snowbitx committed:
test: add test case

snowbitx committed:
Merge branch 'dev' into feat/autocomplete-keydown

tolking reviewed (approved):
LGTM.

rzzf reviewed (approved):
LGTM👍
